File: src/main/management/unix.ts
import { EOL } from 'os';
import path from 'path';
import {
  CLIMAT_HOME_DIR_NAME,
  moveManifestToClimatHome,
  removeAliasSymlinks,
  removeToolchain,
} from './utils';
import fs from 'fs-extra';
import { getInstalledToolchains } from './list';

const join = path.posix.join;
export const CLIMAT_HOME = join(
  '/',
  'usr',
  'local',
  'lib',
  CLIMAT_HOME_DIR_NAME,
);
export const TOOLCHAIN_HOME = join(CLIMAT_HOME, 'toolchains');
const climatScriptBin = join('/', 'usr', 'local', 'bin');

function getScriptContent(name: string): string {
  return (
    '#' + // Need to do this because of the SheBang webpack plugin.
    `!/bin/bash${EOL}${EOL}climat runGlobal "${name}" "$@"${EOL}`
  );
}

export async function unixInstall(
  pathToManifest: string,
  name: string,
  aliases: string[],
): Promise<void> {
  const binPath = join(climatScriptBin, name);
  await moveManifestToClimatHome(TOOLCHAIN_HOME, pathToManifest, name);

  if (!(await fs.pathExists(binPath))) {
    await fs.writeFile(binPath, getScriptContent(name), {
      flag: 'wx',
    });
    await fs.chmod(binPath, 0o755);

    await Promise.all(
      aliases.map(async (alias) => {
        await fs.symlink(binPath, join(climatScriptBin, alias));
        await fs.chmod(binPath, 0o755);
      }),
    );
  }
}

export async function unixUninstall(name: string): Promise<void> {
  await removeAliasSymlinks(name, climatScriptBin, TOOLCHAIN_HOME);
  await removeToolchain(TOOLCHAIN_HOME, name);
  await fs.rm(join(climatScriptBin, name));
}

export async function unixPurge(): Promise<void> {
  await Promise.all(
    (
      await getInstalledToolchains()
    ).map((tc) => fs.rm(join(climatScriptBin, tc))),
  );
  await fs.rm(CLIMAT_HOME, {
    recursive: true,
  });
}

File: src/main/kotlin/management/Unix.kt
package com.climat.management

import com.climat.Fs
import com.climat.jsObjectOf
import kotlinx.coroutines.await
import os.EOL
import path.path
import kotlin.js.Promise

val posix = path.posix
val UNIX_CLIMAT_HOME = posix.join("/", "usr", "local", "lib", CLIMAT_HOME_DIR_NAME)
val UNIX_TOOLCHAIN_HOME = posix.join(UNIX_CLIMAT_HOME, "toolchains")
val climatScriptBin = posix.join("/", "usr", "local", "bin")

fun getScriptContent(name: String) = "#!/bin/bash$EOL${EOL}climat runGlobal \"$name\" \"$@\"$EOL"

suspend fun unixInstall(
    pathToManifest: String,
    name: String,
    aliases: Array<String>,
) {
    val binPath = posix.join(climatScriptBin, name)
    moveManifestToClimatHome(UNIX_TOOLCHAIN_HOME, pathToManifest, name)

    val permissions = 0b111101101 // 755 in octa

    if (!Fs.pathExists(binPath).await()) {
        Fs.writeFile(binPath, getScriptContent(name), "wx").await()
        Fs.chmod(binPath, permissions).await()

        aliases.forEach {
            Fs.symlink(binPath, posix.join(climatScriptBin, it)).await()
            Fs.chmod(binPath, permissions).await()
        }
    }
}

suspend fun unixUninstall(name: String) {
    removeAliasSymlinks(name, climatScriptBin, UNIX_TOOLCHAIN_HOME)
    removeToolchain(UNIX_TOOLCHAIN_HOME, name)
    Fs.rm(posix.join(climatScriptBin, name)).await()
}

suspend fun unixPurge() {
    Promise.all(
        getInstalledToolchains().map {
            Fs.rm(path = posix.join(climatScriptBin, it))
        }.toTypedArray(),
    ).await()

    Fs.rm(
        UNIX_CLIMAT_HOME,
        jsObjectOf(
            "recursive" to true,
        ),
    )
}
